### PR TITLE
IonReader Builder for IonValues

### PR DESCRIPTION
**Pull request by R-maan**
Ongoing

### CODE REVIEW

File: IonDotnet/Internals/Tree/SystemTreeReader.cs [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+        {
+            return _current.Type() == IonType.Struct
+                || _current.Type() == IonType.List
+                || _current.Type() == IonType.Sexp
+                || _current.Type() == IonType.Datagram;
+        }
+
+        private void Push()
+        {
+            int oldlen = _stack.Length;
```
Comment (byronlin13) [inline IonDotnet/Internals/Tree/SystemTreeReader.cs]: camel case for local variables

File: IonDotnet/Internals/Tree/SystemTreeReader.cs [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+using System.Collections.Generic;
+using System.Numerics;
+using System.Text;
+
+namespace IonDotnet.Internals.Tree
+{
+    internal class SystemTreeReader : IIonReader
+    {
+
+        protected readonly IIonValue _value;
```
Comment (byronlin13) [inline IonDotnet/Internals/Tree/SystemTreeReader.cs]: This variable doesn't seem to be used anywhere

File: IonDotnet/Internals/Tree/SystemTreeReader.cs [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+        public virtual ISymbolTable GetSymbolTable() => _systemSymbols;
+
+        public IEnumerable<SymbolToken> GetTypeAnnotations()
+        {
+            return _current.GetTypeAnnotations();
+        }
+
+        public int IntValue()
+        {
+            return _next.IntValue;
```
Comment (byronlin13) [inline IonDotnet/Internals/Tree/SystemTreeReader.cs]: Should this be _current.IntValue?

File: IonDotnet/Internals/Tree/SystemTreeReader.cs [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+                        return _parent.GetElementAt(_next_idx);
+                    }
+                    catch (IndexOutOfRangeException)
+                    {
+                        throw new InvalidOperationException();
+                    }
+                }
+            }
+
+            object IEnumerator.Current => _curr;
```
Comment (byronlin13) [inline IonDotnet/Internals/Tree/SystemTreeReader.cs:281]: What does this do?

File: IonDotnet/Internals/Tree/UserTreeReader.cs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace IonDotnet.Internals.Tree
+{
+    internal class UserTreeReader : SystemTreeReader
+    {
+        /// The ID of system symbol {@value #ION_1_0}, as defined by Ion 1.0.
+        public static readonly int ION_1_0_SID = 2;
```
Comment (byronlin13) [inline IonDotnet/Internals/Tree/UserTreeReader.cs]: Should we use const instead?

File: IonDotnet/Internals/Tree/UserTreeReader.cs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+                }
+                // if we get here we didn't process a system
+                // value, if we had we would have 'continue'd
+                // so this is a value the user gets
+                break;
+            }
+            return (next_type != IonType.Null);
+        }
+
+        private int _symbol_table_top = 0;
```
Comment (byronlin13) [inline IonDotnet/Internals/Tree/UserTreeReader.cs]: Should we group these variables to the top?

Comment (R-maan) [reply IonDotnet/Internals/Tree/UserTreeReader.cs]: Done.

File: IonDotnet/Internals/Tree/UserTreeReader.cs (hunk 5 above)
Comment (R-maan) [reply IonDotnet/Internals/Tree/UserTreeReader.cs]: Done.

File: IonDotnet/Internals/Tree/SystemTreeReader.cs (hunk 4 above)
Comment (R-maan) [reply IonDotnet/Internals/Tree/SystemTreeReader.cs:281]: Implementations of  ```interface IEnumerator<out T>``` should support simple iteration over a generic collection.

File: IonDotnet/Internals/Tree/SystemTreeReader.cs (hunk 3 above)
Comment (R-maan) [reply IonDotnet/Internals/Tree/SystemTreeReader.cs]: That's right, thanks.

File: IonDotnet/Internals/Tree/SystemTreeReader.cs (hunk 1 above)
Comment (R-maan) [reply IonDotnet/Internals/Tree/SystemTreeReader.cs]: Done.

File: IonDotnet/Internals/Tree/SystemTreeReader.cs (hunk 2 above)
Comment (R-maan) [reply IonDotnet/Internals/Tree/SystemTreeReader.cs]: Removed.

**Comment by Armanbqt** [comment]
Merged to amz/mster - updated the fork.

**Comment by R-maan** [comment]
Merged to amz/mster - updated the fork

**Pull request closed without merging**